lisa/tpp_model.py

```diff
@@ -0,0 +1,634 @@
+        if not cur_pe.task:
+            cur_pid = 0
+            cur_comm = 'swapper'
+        else:
+            cur_pid = cur_pe.task.pid
+            cur_comm = cur_pe.task.name
+        if not prev_task:
+            prev_pid = 0
+            prev_comm = 'swapper'
+        else:
+            prev_pid = prev_task.pid
+            prev_comm = prev_task.name
+        if not next_task:
+            next_pid = 0
+            next_comm = 'swapper'
+        else:
+            next_pid = next_task.pid
+            next_comm = next_task.name
+        prev_prio = 100
+        next_task_prio = 100
+        
+        self.buffers['sched_switch'].append((self.unique_time_stamp(), cur_pe.cpuid, cur_pid, cur_comm, prev_comm, prev_pid, prev_prio, prev_state, next_comm, next_pid, next_task_prio))
+
+    def trace_sched_migrate_task(self, prev_pe, next_pe, task):
+        if not prev_pe.task:
+            cur_pid = 0
+            cur_comm = 'swapper'
+        else:
+            cur_pid = prev_pe.task.pid
+            cur_comm = prev_pe.task.name
+        task_prio = 100
+        self.buffers['sched_migrate_task'].append((self.unique_time_stamp(), prev_pe.cpuid, cur_pid, cur_comm, task.name, next_pe.cpuid, prev_pe.cpuid, task.pid, task_prio))
+
+    def trace_cpu_frequency(self, pe, capacity):
+        cur_pid = pe.task.pid
+        cur_comm = pe.task.name
+        freq = int(capacity * 1e6)
+        self.buffers['cpu_frequency'].append((self.unique_time_stamp(), pe.cpuid, cur_pid, cur_comm, pe.cpuid, freq))
+
+    def trace_cpu_idle(self, pe, state):
+        cur_pid = pe.task.pid
+        cur_comm = pe.task.name
+        self.buffers['cpu_idle'].append((self.unique_time_stamp(), pe.cpuid, cur_pid, cur_comm, pe.cpuid, state))
+
+    def trace_sim_activation_complete(self, pe, task, start, end):
+        self.buffers['sim_activation_complete'].append((self.unique_time_stamp(), pe.cpuid, pe.task.pid, pe.task.name, pe.cpuid, task.pid, start, end))
+
+
+class TaskState():
+    TASK_RUNNING = 0x0000
+    TASK_INTERRUPTIBLE = 0x0001
+    TASK_UNINTERRUPTIBLE = 0x002
+
+class SimulationConf(SimpleMultiSrcConf):
+    """
+    YAML configuration helper.
+    """
+    CPUCapacities = TypedDict[int,int]
+    FreqList = SortedTypedList[int]
+    CPUIdList = SortedTypedList[int]
+
+    STRUCTURE = TopLevelKeyDesc('tpp-conf', 'Simulation configuration', (
+        KeyDesc('name', 'Simulation instance name', [str]),
+        LevelKeyDesc('cpu', 'PE properties', (
+            KeyDesc('capacities', 'Compute capacities of CPUs', [CPUCapacities]),
+            KeyDesc('freqs', 'CPU performance levels', [TypedDict[int,FreqList]]),
+            KeyDesc('perf-domains', 'CPU performance domain', [TypedList[CPUIdList]]),
+        )),
+        LevelKeyDesc('workload', 'workload properties', (
+            KeyDesc('tasks', 'tasks (busy_time [s], period [s], offset [s])', [TypedDict[int,TypedList[float]]]),
+        ))
+    ))
+
+class Simulation(Configurable, Loggable):
```

Comment:
We will probably want a `get_trace()` method that would return a `lisa.trace.Trace` loaded with the event buffers and the correct timestamps.